modules/cmpapi/src/encoder/GppModel.ts

```diff
@@ -226,94 +195,100 @@ export class GppModel {
-    if (!this.dirty) {
-      return this.encodedString;
-    }
-
-    // lazily decode
-    if (!this.decoded && this.encodedString != null && this.encodedString.length > 0) {
-      try {
-        this.decode(this.encodedString);
-      } catch (e) {
-        throw new LazyDecodingError(e.message);
-      }
-    }
-
+  protected encodeModel(sections: Map<string, EncodableSection>) {
     let encodedSections = [];
     let sectionIds = [];
     for (let i = 0; i < Sections.SECTION_ORDER.length; i++) {
       let sectionName = Sections.SECTION_ORDER[i];
-      if (this.sections.has(sectionName)) {
-        let section = this.sections.get(sectionName);
+      if (sections.has(sectionName)) {
+        let section = sections.get(sectionName);
         encodedSections.push(section.encode());
         sectionIds.push(section.getId());
       }
     }
 
     let header = new HeaderV1();
-    header.setFieldValue("SectionIds", this.getSectionIds());
+    header.setFieldValue("SectionIds", sectionIds);
     encodedSections.unshift(header.encode());
 
-    this.encodedString = encodedSections.join("~");
-    this.dirty = false;
-    return this.encodedString;
+    return encodedSections.join("~");
   }
 
-  public decode(str: string) {
-    this.encodedString = str;
-    this.decoded = false;
-    this.dirty = true;
-    this.sections.clear();
+  protected decodeModel(str: string) {
+    if (!str || str.length == 0 || str.startsWith("DB")) {
+      let encodedSections = str.split("~");
+      let sections = new Map<string, EncodableSection>();
+      if (encodedSections[0].startsWith("D")) {
+        //GPP String
+        let header = new HeaderV1(encodedSections[0]);
+        let sectionIds = header.getFieldValue("SectionIds");
+
+        if (sectionIds.length !== encodedSections.length - 1) {
+          throw new DecodingError(
+            "Unable to decode '" +
+              str +
+              "'. The number of sections does not match the number of sections defined in the header."
+          );
+        }
+
+        for (let i = 0; i < sectionIds.length; i++) {
+          let encodedSection = encodedSections[i + 1];
+          if (encodedSection.trim() === "") {
+            throw new DecodingError("Unable to decode '" + str + "'. Section " + (i + 1) + " is blank.");
+          }
+          if (sectionIds[i] === TcfCaV1.ID) {
+            let section = new TcfCaV1(encodedSection);
+            sections.set(TcfCaV1.NAME, section);
+          } else if (sectionIds[i] === TcfEuV2.ID) {
+            let section = new TcfEuV2(encodedSection);
+            sections.set(TcfEuV2.NAME, section);
+          } else if (sectionIds[i] === UspV1.ID) {
+            let section = new UspV1(encodedSection);
+            sections.set(UspV1.NAME, section);
+          } else if (sectionIds[i] === UsNatV1.ID) {
+            let section = new UsNatV1(encodedSection);
+            sections.set(UsNatV1.NAME, section);
+          } else if (sectionIds[i] === UsCaV1.ID) {
```

Comment:
as of #54 all these v1 need to be removed

Reply:
I'll address this in a separate PR.